### PR TITLE
Integrate Namespaces with Player

### PR DESCRIPTION
**Pull request by jackrieck**
# Summary

Integrate `Player` and `PlayerClass` accounts with the `Namespace` program.

Includes modifications to `Player` and `Namespace` Programs, both will need to be upgraded on chain. Tests included proving it works

### CODE REVIEW

**Review by long-banana — CHANGES_REQUESTED** [review]
Update a couple comments if what I requested makes sense. Besides that change, this is 🔥

File: js/lib/src/instructions/player.ts [hunk 1]
```diff
@@ -164,7 +164,7 @@ export interface CreatePlayerClassArgs {
   parentOfParentClassIndex: null | BN;
   parentClassIndex: null | BN;
   space: BN;
-  desiredNamespaceArraySize: number;
+  desiredNamespaceArraySize: BN;
```
[the remark this answers is not in the document]
Comment (jackrieck) [reply js/lib/src/instructions/player.ts:167]: Good call, gonna revert for now, ran the tests without it and I dont need to make the change

File: rust/player/src/lib.rs [hunk 2]
```diff
@@ -1837,6 +1961,63 @@ pub struct UpdatePlayer<'info> {
     // See the [COMMON REMAINING ACCOUNTS] ctrl f for this
 }
 
+#[derive(Accounts)]
+pub struct PlayerArtifactJoinNamespace<'info> {
+    /// CHECK: deserialized inside instruction
```
[the remark this answers is not in the document]
Comment (jackrieck) [reply rust/player/src/lib.rs]: Added more comments about the flow

**Review by long-banana — APPROVED** [review]

File: rust/matches/src/utils.rs [hunk 3]
```diff
@@ -315,7 +315,7 @@ pub fn verify(proof: &Vec<[u8; 32]>, root: &[u8; 32], leaf: [u8; 32]) -> bool {
 pub fn is_part_of_namespace<'a>(artifact: &AccountInfo<'a>, namespace: &Pubkey) -> bool {
     let data = artifact.data.borrow_mut();
     let number = u32::from_le_bytes(*array_ref![data, 9, 4]) as usize;
-    let offset = 13 as usize;
+    let offset = 13_usize;
```
Comment (long-banana) [inline rust/matches/src/utils.rs:318]: I've never seen this before. How does that work?

File: rust/player/src/utils.rs [hunk 4]
```diff
@@ -2172,5 +2166,53 @@ pub fn is_namespace_program_caller(ixns: &AccountInfo) -> bool {
...
     };
 
-    return true;
+    true
+}
+
+/// Create account almost from scratch, lifted from
+/// https://github.com/solana-labs/solana-program-library/tree/master/associated-token-account/program/src/processor.rs#L51-L98
+#[inline(always)]
+pub fn create_or_allocate_account_raw<'a>(
```
Comment (long-banana) [inline rust/player/src/utils.rs:2175]: What is this for?

**Review by long-banana — APPROVED** [review]
I had one question I'd like you to take a look at before merging about the raw account creation function.

File: rust/matches/src/utils.rs (hunk 3 above)
Comment (jackrieck) [reply rust/matches/src/utils.rs:318]: apparently this is with the new anchor 0.25.0 linting

File: rust/player/src/utils.rs (hunk 4 above)
Comment (jackrieck) [reply rust/player/src/utils.rs:2175]: Im not sure why its there, let me double check

**Pull request merged**